### PR TITLE
No ethabi

### PR DESCRIPTION
**Pull request by MrChico**
Reimplements `seth calldata`, `seth --abi-decode` and `seth --decode-event` to use ethers.js instead of ethabi, removing the ethabi dependency completely.
solves https://github.com/dapphub/dapptools/issues/298

### CODE REVIEW

File: src/seth/libexec/seth/seth-calldata [hunk 1]
```diff
@@ -1,35 +1,29 @@
...
+    const funcs = new ethers.utils.Interface(["function " + sig.replace(')(', ') returns (')]).functions;
+      console.log(funcs[Object.keys(funcs)[0]].encode(process.argv.slice(3)))
+    } catch (e) {
+      console.log(e.toString())
+      process.exit(1)
+    }
+} else if (process.argv.length == 3) {
+  const {execSync}   = require('child_process');
+  try {
+    const yes = execSync('seth --to-hexdata \"' + process.argv[2] + "\"").toString().replace('\n','')
```
[the remark this answers is not in the document]
Comment (asymmetric) [reply src/seth/libexec/seth/seth-calldata]: Also, I'd use [`execFileSync`](https://nodejs.org/api/child_process.html#child_process_child_process_execfilesync_file_args_options) over `execSync`, as by doing so we don't have to think about arbitrary shell code execution.

File: src/dapp/Dockerfile [hunk 2]
```diff
@@ -22,7 +22,6 @@ ENV USER node
 WORKDIR /home/node
 
 RUN curl https://nixos.org/nix/install | bash
-RUN /nix/var/nix/profiles/default/bin/nix-env -i ethabi jshon
```
[the remark this answers is not in the document]
Comment (MrChico) [reply src/dapp/Dockerfile]: oops. I'll put jshon back

File: src/seth/libexec/seth/seth-calldata (hunk 1 above)
[the remark this answers is not in the document]
Comment (MrChico) [reply src/seth/libexec/seth/seth-calldata]: sig is out of scope here

**Comment by MrChico** [comment]
@asymmetric thanks for the review.

> @MrChico I've done some very cursory testing and the results look the same. Have you checked that as well?

I've been careful to maintain the semantics of `seth calldata` and `seth ---abi-decode`. `--decode-event` I'm not as sure about but I doubt that it is even being used by anyone...

I've spent some time doing what you might call "integration tests" for this by using this fork in a project (where the issues with `ethabi` originally arose), and it solves the issues I was experiencing. There are some differences, in that `etherjs` seems better typed which has led me to include the changes from https://github.com/dapphub/dapptools/pull/315 here as well.

**Comment by MrChico** [comment]
Any idea why ci is failing @asymmetric ? Travis seem to have start some form of monologue before erroring out with `log limit reached`

**Comment by asymmetric** [comment]
@MrChico it's just a timeout, I've restarted the job.

**Comment by asymmetric** [comment]
LOL sorry, should've read the message more carefully. It's due to [this policy](https://docs.travis-ci.com/user/common-build-problems/#log-length-exceeded).

**Comment by asymmetric** [comment]
[These lines](https://travis-ci.com/dapphub/dapptools/jobs/277422042#L230-L235) are very weird btw.

**Comment by MrChico** [comment]
Yeah. But that issue does not seem related to the latest change.

**Comment by asymmetric** [comment]
Nope. I suspect the whole CI pipeline is broken for some reason.

**Comment by asymmetric** [comment]
Rebased on top of #332.

**Comment by MrChico** [comment]
FYI I've been using this quite a bit over the last couple of weeks and haven't had any problems. Now with ci passing this is ready to merge as far as I'm concerned

**Comment by MrChico** [comment]
I'm using this in production and will merge in a few days unless anyone expresses dissent

**Comment by asymmetric** [comment]
@MrChico I don't have enough exposure to the tooling day-to-day to evaluate whether this is working as intended.

Ideally we'd have some tests, not sure if you want to block this PR on adding them, but it it does make me slightly uneasy that we don't have any, especially when making these types of changes.

**Comment by MrChico** [comment]
agree

**Pull request merged**